### PR TITLE
Handle MACOSX versions >10 or with 2 dots

### LINKED ISSUE #1238: MACOSX_DEPLOYMENT_TARGET misconfiguration on macOS 11.0.1
It appears that luarocks misconfigures the `CC` and `LD` variables on macOS 11.0.1 resulting in the following warnings and potentially other problems:

```
$ luarocks --local install luacheck
Installing https://luarocks.org/luacheck-0.24.0-2.src.rock
Missing dependencies for luacheck 0.24.0-2:
   argparse >= 0.6.0 (not installed)
   luafilesystem >= 1.6.3 (not installed)

luacheck 0.24.0-2 depends on argparse >= 0.6.0 (not installed)
Installing https://luarocks.org/argparse-0.7.1-1.src.rock
No existing manifest. Attempting to rebuild...
argparse 0.7.1-1 is now installed in /Users/jenders/.luarocks (license: MIT)

luacheck 0.24.0-2 depends on luafilesystem >= 1.6.3 (not installed)
Installing https://luarocks.org/luafilesystem-1.8.0-1.src.rock
env MACOSX_DEPLOYMENT_TARGET=10.0 /usr/bin/clang -O2 -fPIC -I/opt/local/include -c src/lfs.c -o src/lfs.o
env MACOSX_DEPLOYMENT_TARGET=10.0 /usr/bin/clang -bundle -undefined dynamic_lookup -all_load -o lfs.so -L/opt/local/lib src/lfs.o
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/bundle1.o) was built for newer macOS version (10.4) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(comparetf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(extenddftf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(extendsftf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(fixtfdi.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(fixtfsi.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(fixtfti.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(fixunstfdi.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(fixunstfsi.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(fixunstfti.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(floatditf.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(floatsitf.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(floattitf.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(floatunditf.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(floatunsitf.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(floatuntitf.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(multc3.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(trunctfdf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(trunctfsf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(bswapdi2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(bswapsi2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(divtc3.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(extendhfsf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(ffssi2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(int_util.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(mulodi4.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(mulosi4.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(muloti4.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(os_version_check.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(truncdfhf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(truncsfhf2.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(emutls.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(atomic_flag_clear.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(atomic_flag_clear_explicit.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(atomic_flag_test_and_set.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(atomic_flag_test_and_set_explicit.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(atomic_signal_fence.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(atomic_thread_fence.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(chkstk_darwin.S.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(chkstk_darwin2.S.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(cpu_model.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
ld: warning: object file (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.osx.a(fp_mode.c.o)) was built for newer macOS version (10.5) than being linked (10.0)
luafilesystem 1.8.0-1 is now installed in /Users/jenders/.luarocks (license: MIT/X11)

luacheck 0.24.0-2 is now installed in /Users/jenders/.luarocks (license: MIT)
```

This appears to be due to the fact that the pattern used to match the version returned by `sw_vers` in [src/luarocks/core/cfg.lua](https://github.com/luarocks/luarocks/blob/d2c3b57b324f06d17cd2c594e26b143dfe3a8133/src/luarocks/core/cfg.lua#L454-L471) only matches the minor version which of course "rolls over" on 11.0.1:

```lua
   if platforms.macosx then
      defaults.variables.MAKE = "make"
      defaults.external_lib_extension = "dylib"
      defaults.arch = "macosx-"..target_cpu
      defaults.variables.LIBFLAG = "-bundle -undefined dynamic_lookup -all_load"
      local version = util.popen_read("sw_vers -productVersion")
      version = tonumber(version and version:match("^[^.]+%.([^.]+)")) or 3
      if version >= 10 then
         version = 8
      elseif version >= 5 then
         version = 5
      else
         defaults.gcc_rpath = false
      end
      defaults.variables.CC = "env MACOSX_DEPLOYMENT_TARGET=10."..version.." gcc"
      defaults.variables.LD = "env MACOSX_DEPLOYMENT_TARGET=10."..version.." gcc"
      defaults.web_browser = "open"
   end
```

This mistake is so common that you can actually pass `SYSTEM_VERSION_COMPAT` to `sw_vers` to produce a false minor version but I don't think this is the right solution here.

```
$ sw_vers -productVersion
11.0.1
$ SYSTEM_VERSION_COMPAT=1 sw_vers -productVersion
10.16
```

I'd like to submit a PR but I can think of several ways to fix this and guidance from experienced project contributors  for the preferred method would be appreciated.

### PR DESCRIPTION
**Pull request by sharpobject**
Fixes #1238.

I had trouble installing cqueues so I thought I would fix it: https://github.com/wahern/cqueues/issues/246

This patch is what @siffiejoe recommended here: https://github.com/luarocks/luarocks/issues/1238#issuecomment-727590010

### CODE REVIEW

File: src/luarocks/build.lua [hunk 1]
```diff
@@ -94,10 +92,12 @@ local function check_macosx_deployment_target(rockspec)
...
       local version = util.popen_read("sw_vers -productVersion")
-      local versionminor = minor(version)
-      local targetminor = minor(target)
-      if targetminor > versionminor then
-         return nil, ("This rock requires Mac OSX 10.%d, and you are running 10.%d."):format(targetminor, versionminor)
+      if version:match("^%d+%.%d+%.%d+$") or version:match("^%d+%.%d+$") then
+         version = vers.parse_version(version)
+         local targetversion = vers.parse_version(target)
+         if targetversion > version then
+            return nil, ("This rock requires Mac OSX %s, and you are running %s."):format(targetversion, version)
```
Comment (daurnimator) [inline src/luarocks/build.lua:99]: you might need `tostring` here; lua 5.1 doesn't call it for us on string.format arguments.

File: src/luarocks/build.lua [hunk 2]
```diff
@@ -12,6 +12,7 @@ local cfg = require("luarocks.core.cfg")
 local repos = require("luarocks.repos")
 local writer = require("luarocks.manif.writer")
 local deplocks = require("luarocks.deplocks")
+local vers = require("luarocks.core.vers")
```
Comment (daurnimator) [inline src/luarocks/build.lua:15]: nit: could you move this up next to the other `core` module required?

**Review by daurnimator — CHANGES_REQUESTED** [review]
LGTM, can you test with all lua versions on OSX?
I don't have an OSX machine to test with.

File: src/luarocks/build.lua (hunk 2 above)
Comment (sharpobject) [reply src/luarocks/build.lua:15]: Alright!

File: src/luarocks/build.lua (hunk 1 above)
Comment (sharpobject) [reply src/luarocks/build.lua:99]: You're right!

**Comment by sharpobject** [comment]
Seems to work on all lua versions under Mac OS 11.1:
```
#!/bin/bash

set -e

for VERSION in 5.1.0 5.1.1 5.1.2 5.1.3 5.1.4 5.1.5 5.2.0 5.2.1 5.2.2 5.2.3 5.2.4 5.3.0 5.3.1 5.3.2 5.3.3 5.3.4 5.3.5 5.3.6 5.4.0 5.4.1 5.4.2
do
  hererocks lua$VERSION --lua $VERSION --luarocks https://github.com/sharpobject/luarocks.git@master
  source lua$VERSION/bin/activate
  luarocks install cqueues CRYPTO_DIR=/opt/homebrew/opt/openssl OPENSSL_DIR=/opt/homebrew/opt/openssl
done

# unsupported on arm64: 2.0.0 2.0.1 2.0.2 2.0.3 2.0.4 2.0.5

for VERSION in 2.1.0-beta1 2.1.0-beta2 2.1.0-beta3
do
  hererocks luajit$VERSION --luajit $VERSION --luarocks https://github.com/sharpobject/luarocks.git@master
  source luajit$VERSION/bin/activate
  luarocks install cqueues CRYPTO_DIR=/opt/homebrew/opt/openssl OPENSSL_DIR=/opt/homebrew/opt/openssl
done

export MACOSX_DEPLOYMENT_TARGET=10.8
hererocks luajit-agentzh --luajit https://github.com/openresty/luajit2.git@v2.1-agentzh --luarocks https://github.com/sharpobject/luarocks.git@master
source luajit-agentzh/bin/activate
luarocks install cqueues CRYPTO_DIR=/opt/homebrew/opt/openssl OPENSSL_DIR=/opt/homebrew/opt/openssl
```

I did not try it on luajit 2.0.x under Rosetta 2 because of laziness.

**Review by daurnimator — APPROVED** [review]

**Pull request merged**